### PR TITLE
Move toolstate checking into bootstrap

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This intends no functional changes, merely translates the spread of shell/python into Rust.

One problematic area that I'd like to avoid but wasn't quite able to figure out how is the master branch script which is still in bash/python -- I cared less about that since it is orthogonal to the actual checking that we're doing, though as-is we're duplicating some code across Rust and that script.

r? @kennytm or maybe @pietroalbini

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
This is preparation for https://github.com/rust-lang/rust/issues/65000 -- but *not* the changes that the FCP there is about; that will be done in a separate, future PR.

**Comment by Mark-Simulacrum** [comment]
(CI failure is expected and due to the last commit which tests that we're still verifying submodules)

**Comment by Mark-Simulacrum** [comment]
(If we're down with the approach the rebase and such here should be mostly painless)

File: src/bootstrap/toolstate.rs [hunk 1]
```diff
@@ -1,4 +1,21 @@
...
+use crate::builder::{Builder, RunConfig, ShouldRun, Step};
+use std::fmt;
+use std::process::Command;
+use std::path::PathBuf;
+use std::env;
+
+#[cfg(unix)]
+const OS: &str = "linux";
+#[cfg(not(unix))]
+const OS: &str = "windows";
```
Comment (pietroalbini) [inline src/bootstrap/toolstate.rs]: Can we have this? I don't want macOS to be treated as Linux if we (accidentally or not) run toolstate on it.

```rust
#[cfg(linux)]
const OS: Option<&str> = Some("linux");

#[cfg(windows)]
const OS: Option<&str> = Some("windows");

#[cfg(all(not(linux), not(windows)))]
const OS: Option<&str> = None;
```

File: src/bootstrap/toolstate.rs [hunk 2]
```diff
@@ -12,9 +29,390 @@ pub enum ToolState {
...
     }
 }
+
+/// Number of days after the last promotion of beta.
+/// Its value is 41 on the Tuesday where "Promote master to beta (T-2)" happens.
+/// The Wednesday after this has value 0.
+/// We track this value to prevent regressing tools in the last week of the 6-week cycle.
+fn days_since_beta_promotion() -> u64 {
+    let since_epoch = t!(time::SystemTime::UNIX_EPOCH.elapsed());
+    (since_epoch.as_secs() / 86400 - 20) % 42
```
Comment (pietroalbini) [inline src/bootstrap/toolstate.rs]: The `42` is a bit hidden, I'd extract it into a constant at the top so we can easily tweak it in the future.

File: src/bootstrap/toolstate.rs [hunk 3]
```diff
@@ -12,9 +29,390 @@ pub enum ToolState {
...
+    ///
+    /// This tool in `src/tools` will verify the validity of all our links in the
+    /// documentation to ensure we don't have a bunch of dead ones.
+    fn run(self, builder: &Builder<'_>) {
+        if builder.config.dry_run {
+            return;
+        }
+
+        let days_since_beta_promotion = days_since_beta_promotion();
+        let in_beta_week = days_since_beta_promotion >= 35;
```
Comment (pietroalbini) [inline src/bootstrap/toolstate.rs]: The `35` is a bit hidden, I'd extract it into a constant at the top so we can easily tweak it in the future.

File: src/bootstrap/toolstate.rs [hunk 4]
```diff
@@ -12,9 +29,390 @@ pub enum ToolState {
...
+///
+///   2. Update the variable group in Azure Pipelines
+///
+///       * Ping a member of the infrastructure team to do this.
+///
+///   4. Replace the email address below if the bot account identity is changed
+///
+///       * See <https://help.github.com/articles/about-commit-email-addresses/>
+///           if a private email by GitHub is wanted.
+fn commit_toolstate_change(
```
Comment (pietroalbini) [inline src/bootstrap/toolstate.rs]: Can we switch to use deploy keys instead? Dunno if you want to do this in this PR or another, but it's something I think we should do eventually. The code to push with a deploy key is [here](https://github.com/rust-lang/simpleinfra/blob/master/setup-deploy-keys/src/deploy.rs).

File: src/bootstrap/toolstate.rs [hunk 5]
```diff
@@ -12,9 +29,390 @@ pub enum ToolState {
...
+
+fn change_toolstate(
+    current_toolstate: &ToolstateData,
+    old_toolstate: &[RepoState],
+    in_beta_week: bool,
+) {
+    let mut regressed = false;
+    for repo_state in old_toolstate {
+        let tool = &repo_state.tool;
+        let state = if cfg!(unix) {
```
Comment (pietroalbini) [inline src/bootstrap/toolstate.rs]: Same as above, I'd do `cfg!(linux)`, `cfg!(windows)` and then an `unimplemented!()`.

**Review by emilyalbini — DISMISSED** [review]
> One problematic area that I'd like to avoid but wasn't quite able to figure out how is the master branch script which is still in bash/python -- I cared less about that since it is orthogonal to the actual checking that we're doing, though as-is we're duplicating some code across Rust and that script.

I don't think it's that bad if we initialize the full CI environment in the master builder (it's like a couple of minutes?). Then we can just do `./x.py test publish-toolstate` (switching to `run` once my GHA PR lands).

File: src/bootstrap/toolstate.rs (hunk 2 above)
Comment (Mark-Simulacrum) [reply src/bootstrap/toolstate.rs]: The 42 is just the number of days in a single cycle, which seems unlikely to change (certainly not often)?

File: src/bootstrap/toolstate.rs (hunk 4 above)
Comment (Mark-Simulacrum) [reply src/bootstrap/toolstate.rs]: I'd rather do this in a separate PR, seems cleaner.

**Comment by Mark-Simulacrum** [comment]
> I don't think it's that bad if we initialize the full CI environment in the master builder (it's like a couple of minutes?). Then we can just do ./x.py test publish-toolstate (switching to run once my GHA PR lands).

Yeah, this seems reasonable. I would prefer that we land this and then do that in a followup PR -- I can work on it pretty quickly (mostly because I worry about changing too much in one PR, and the existing duplication is pretty minimal).

I believe I've resolve the other review comments you left.

**Comment by pietroalbini** [comment]
@bors r+ rollup=never

Thanks!

**Pull request merged**